staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go

```diff
@@ -46,42 +49,43 @@ type timeBudget interface {
 
 type timeBudgetImpl struct {
 	sync.Mutex
-	budget time.Duration
-
-	refresh   time.Duration
+	clock clock.Clock
+	// budget is the amount of accumulated time
+	budget    time.Duration
 	maxBudget time.Duration
+	// each refreshInterval add a refresh amount of time to the budget
+	refresh         time.Duration
+	refreshInterval time.Duration
+	// last stores the times the timeBudget has been accessed
+	last time.Time
 }
 
-func newTimeBudget(stopCh <-chan struct{}) timeBudget {
+func newTimeBudget() timeBudget {
 	result := &timeBudgetImpl{
-		budget:    time.Duration(0),
-		refresh:   refreshPerSecond,
-		maxBudget: maxBudget,
+		clock:           clock.RealClock{},
+		budget:          time.Duration(0),
+		refresh:         refreshPerInterval,
+		refreshInterval: refreshInterval,
+		maxBudget:       maxBudget,
 	}
-	go result.periodicallyRefresh(stopCh)
+	result.last = result.clock.Now()
 	return result
 }
 
-func (t *timeBudgetImpl) periodicallyRefresh(stopCh <-chan struct{}) {
-	ticker := time.NewTicker(time.Second)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ticker.C:
-			t.Lock()
-			if t.budget = t.budget + t.refresh; t.budget > t.maxBudget {
-				t.budget = t.maxBudget
-			}
-			t.Unlock()
-		case <-stopCh:
-			return
-		}
-	}
-}
-
 func (t *timeBudgetImpl) takeAvailable() time.Duration {
 	t.Lock()
 	defer t.Unlock()
+	// budget accumulated since last access
+	acc := t.clock.Since(t.last).Seconds() * (t.refresh.Seconds() / t.refreshInterval.Seconds())
+	// this shouldn't be possible
+	if acc < 0 {
+		return time.Duration(0)
```

Comment:
This isn't correct - we should return t.budget, right?

What you need here is to simply set acc = 0 in this case and proceed.

[BTW - a unit test for this case would be useful]